### PR TITLE
Abstract function for the classifier pipeline operators

### PR DESCRIPTION
**Pull request by bartleyn**
Per #43 , this abstracts the shared code between static models into a static method that each model can call. In addition, I built a test for testing all the static models / this new abstract method indirectly, and tweaked the num_trees parameter in random_forest to n_estimators to facilitate testing. 

Potential improvements:
1) Maybe it's just me, but It feels a little odd to call a static function from another static function. Does it make sense to move this method (and perhaps some others) to a utility function module? Either way, is it faster to call this abstract method as a static method inside the class, a class method inside the class, or something else?
2) Is there a more appropriate name for the method?
3) Is there a better place to do the input validation (i.e. if there are only three columns in the input_df)  than within this nested function call?

### CODE REVIEW

File: tpot/tpot.py [hunk 1]
```diff
@@ -724,41 +724,17 @@ def decision_tree(input_df, max_features, max_depth):
...
-        input_df.loc[:, 'guess'] = dtc.predict(all_features)
-
-        # Also store the guesses as a synthetic feature
-        sf_hash = '-'.join(sorted(input_df.columns.values))
-        sf_hash += 'DT-{}-{}'.format(max_features, max_depth)
-        sf_identifier = 'SyntheticFeature-{}'.format(hashlib.sha224(sf_hash.encode('UTF-8')).hexdigest())
-        input_df.loc[:, sf_identifier] = input_df['guess'].values
-
-        return input_df
+        return TPOT._train_model_and_predict(input_df, DecisionTreeClassifier, max_features=max_features, max_depth=max_depth, random_state=42)
```
Comment (rhiever) [inline tpot/tpot.py]: I think the `TPOT.` is extraneous: It should be possible to call `_train_model_and_predict()` without it.

File: tpot/tpot.py [hunk 2]
```diff
@@ -724,41 +724,17 @@ def decision_tree(input_df, max_features, max_depth):
...
-    def random_forest(input_df, num_trees, max_features):
+    def random_forest(input_df, n_estimators, max_features):
         """Fits a random forest classifier
 
         Parameters
         ----------
         input_df: pandas.DataFrame {n_samples, n_features+['class', 'group', 'guess']}
             Input DataFrame for fitting the random forest
-        num_trees: int
+        n_estimators: int
```
Comment (rhiever) [inline tpot/tpot.py]: Fine with me. :+1:

File: tpot/tpot.py [hunk 3]
```diff
@@ -979,31 +868,63 @@ def gradient_boosting(input_df, learning_rate, n_estimators, max_depth):
...
-        
+
         input_df = input_df.copy()
 
         training_features = input_df.loc[input_df['group'] == 'training'].drop(['class', 'group', 'guess'], axis=1).values
         training_classes = input_df.loc[input_df['group'] == 'training', 'class'].values
+       
+        # Try to seed the random_state parameter if the model accepts it.
+        try:
+            clf = model(random_state=42,**kwargs)
```
Comment (rhiever) [inline tpot/tpot.py]: I'm pondering if seeding the random state is even necessary. Maybe it's even harmful in the long run?

This isn't something we need to address here, but I've been considering exploring that more.

**Comment by rhiever** [comment]
Great work as always! Thank you for adding the test.

Addressing your questions:

>1) Maybe it's just me, but It feels a little odd to call a static function from another static function. Does it make sense to move this method (and perhaps some others) to a utility function module? Either way, is it faster to call this abstract method as a static method inside the class, a class method inside the class, or something else?

AFAIK, there is no performance difference between the two. The difference between a static and regular class function is whether the function relies on the current state of the object, i.e., it affects the scope. Static functions always have their own scope, whereas class functions share the scope of the object they're called on.

I think we have to ask ourselves: Do we want users using these functions outside of TPOT? e.g.,

```python
from tpot import random_forest

result = random_forest(...)
...
```

I originally envisioned the `export()` function doing something like that, but I ended up exporting the pipelines directly to sklearn code (even if it looks ugly). Perhaps we should change all of the functions to regular class methods since we don't really want users doing that. I don't think any of the functions in TPOT were really made to be used outside of TPOT.

>2) Is there a more appropriate name for the method?

Name looks fine to me!

>3) Is there a better place to do the input validation (i.e. if there are only three columns in the input_df) than within this nested function call?

That's a tough one, since we're never really sure where the pipeline operator functions will be getting their input from and where they're sending it to. I think putting in that quick check at the beginning of the function is the best we can do.

**Comment by rhiever** [comment]
IMO we should keep `_train_model_and_predict()` within the TPOT class. It's simply an abstraction of the  TPOT classifier functions.

**Comment by bartleyn** [comment]
I settled on moving `_train_model_and_predict()` out of the class to see if we could remove the `TPOT.` that prefaced the function call in the static models; I agree that ideally the function should remain in the TPOT class. However, without passing an instantiated TPOT object into the static model call, I'm not sure how to avoid the static `TPOT._train_model_and_predict()`.

**Comment by rhiever** [comment]
I see. We should definitely remove all of the `@staticfunction` decorators then, and add `self` as the first parameter to these functions. That will allow you to call the internal functions as `self.function_name()`.

**Comment by bartleyn** [comment]
I'll try to squash some of my commits for brevity's sake.

**Comment by rhiever** [comment]
This looks ready to merge now. Anything else you planning on adding to this PR?

**Comment by bartleyn** [comment]
I think I'm good for now. If you think it's ready, then great!

**Pull request merged**